perf/pkeyread.c

```diff
@@ -0,0 +1,315 @@
+     */
+    for (i = 0; i < NUM_CALLS_PER_TEST / threadcount; i++) {
+        key = PEM_read_bio_PrivateKey(pem, NULL, NULL, NULL);
+        if (key == NULL) {
+            fprintf(stderr, "Failed to create key: %d [%s PEM]\n", i,
+                    sample_names[sample_id]);
+            err = 1;
+            BIO_free(pem);
+            return;
+        }
+        EVP_PKEY_free(key);
+        BIO_reset(pem);
+    }
+
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+
+    BIO_free(pem);
+}
+
+void do_derread(size_t num)
+{
+    const char *keydata;
+    size_t keydata_sz;
+    const unsigned char *pk_buf = NULL;
+    int pk_buf_len;
+    BIO *der;
+    PKCS8_PRIV_KEY_INFO *pkey_info8 = NULL;
+    EVP_PKEY *pkey = NULL;
+    X509_PUBKEY *key = NULL;
+    int i;
+    OSSL_TIME start, end;
+
+    if (sample_id >= SAMPLE_ALL) {
+        fprintf(stderr, "%s no sample key set for test\n", __func__);
+        err = 1;
+        return;
+    }
+
+    keydata = sample_keys[sample_id][FORMAT_DER];
+    keydata_sz = sample_key_sizes[sample_id][FORMAT_DER];
+    der = BIO_new_mem_buf(keydata, keydata_sz);
+    if (der == NULL) {
+        fprintf(stderr, "%s Cannot create mem BIO [%s DER]\n",
+                __func__, sample_names[sample_id]);
+        err = 1;
+        return;
+    }
+
+    start = ossl_time_now();
+
+    for (i = 0; i < NUM_CALLS_PER_TEST / threadcount && err == 0; i++) {
+        pkey_info8 = d2i_PKCS8_PRIV_KEY_INFO_bio(der, &pkey_info8);
+        if (pkey_info8 == NULL) {
+            fprintf(stderr, "%s pkey_info8 is NULL [%s BER]\n",
+                    __func__, sample_names[sample_id]);
+            err = 1;
+            goto error;
+        }
+
+        pkey = EVP_PKCS82PKEY(pkey_info8);
+        if (pkey == NULL) {
+            fprintf(stderr, "%s pkey is NULL [%s BER]\n",
+                    __func__, sample_names[sample_id]);
+            err = 1;
+            goto error;
+        }
+error:
+        BIO_reset(der);
+        X509_PUBKEY_free(key);
+        key = NULL;
+        EVP_PKEY_free(pkey);
+        pkey = NULL;
+        PKCS8_PRIV_KEY_INFO_free(pkey_info8);
+        pkey_info8 = NULL;
+    }
+
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+
+    BIO_free(der);
+}
+
+static int sample_name_to_id(const char *sample_name)
+{
+    int i = 0;
+
+    while (sample_names[i] != NULL) {
+        if (strcasecmp(sample_names[i], sample_name) == 0)
+            break;
+        i++;
+    }
+
+    return i;
+}
+
+static int format_name_to_id(const char *format_name)
+{
+    int i = 0;
+
+    while (format_names[i] != NULL) {
+        if (strcasecmp(format_names[i], format_name) == 0)
+            break;
+        i++;
+    }
+
+    return i;
+}
+
+double get_avcalltime(void)
+{
+    int i;
+    OSSL_TIME us;
+    double avcalltime;
+
+    for (i = 1; i < threadcount; i++)
+        us = ossl_time_add(us, times[i]);
+    us = ossl_time_divide(us, NUM_CALLS_PER_TEST);
```

Comment:
I would do this divide as a "double" type as well.

Reply:
I'm just curious: why do you prefer to use double here instead of keeping operations with integers? I took those changes from Niel, so therefore I'm asking. I somehow feel it's better to keep data as ints and convert to float in very last step. This is just my feeling.

This is the Niel's change I took that from: https://github.com/openssl/tools/pull/184

Reply:
I've since updated my PR in which this change was made, in response to a similar comment @mattcaswell made in #184 .  The long and the short of it is:
1) ossl_time_divide does integer division, and so when the divisor is 'large' we loose information in this operation.  for instance, NUM_CALLS_PER_TEST here is 1x10^5, so we could loose up to 99999 ticks in this operation, or about .1us worth of accuracy.  Casting to a double preserves that
2) the name us formerly referred to an OSSL_TIME struct which was going to be represented as microseconds, but the us variable in the new code is really just tracking a total number of ticks during the execution of the test.  And so the name ttime (total time) seemed more appropriate.

Apologies, I should have given you  a heads up about this.  But if you look at the last fixup commit on #184 you can see the changes I made there in response to matts comment

Reply:
Thank you Niel for clarification. Indeed I've missed this update in your #184 PR which fixes stats.